man/man8/pi.8
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "PUPPET\-DESCRIBE" "8" "February 2011" "Puppet Labs, LLC" "Puppet manual"
.
.SH "NAME"
\fBpuppet\-describe\fR \- Display help about resource types
.
.SH "SYNOPSIS"
Prints help about Puppet resource types, providers, and metaparameters\.
.
.SH "USAGE"
puppet describe [\-h|\-\-help] [\-s|\-\-short] [\-p|\-\-providers] [\-l|\-\-list] [\-m|\-\-meta]
.
.SH "OPTIONS"
.
.TP
\-\-help
Print this help text
.
.TP
\-\-providers
Describe providers in detail for each type
.
.TP
\-\-list
List all types
.
.TP
\-\-meta
List all metaparameters
.
.TP
\-\-short
List only parameters without detail
.
.SH "EXAMPLE"
.
.nf

$ puppet describe \-\-list
$ puppet describe file \-\-providers
$ puppet describe user \-s \-m
.
.fi
.
.SH "AUTHOR"
David Lutterkort
.
.SH "COPYRIGHT"
Copyright (c) 2005 Puppet Labs, LLC Licensed under the GNU Public License


CHANGELOG
2.6.5rc5
========
b481321 (#6337) Fix Ruby warning on 1.8.6 about "future compatibility"
dcce45c (#6353) Restore the ability to store paths in the filebucket
0450197 (#6126) Puppet inspect now reports status after run completes.
960890f (#6364) Adjust mis-translated regex in mount provider for AIX

2.6.5rc4
========
664ef67 (#3646) Fix the documentation fix for `puppet apply --apply`

2.6.5rc3
========
7ef2fbf Updated fix for #3646 - apply / compile documentation
193016d (#5977) fix spec test failure when new applications are introduced.

2.6.5rc2
========
1f89906 (#6257) Speed up PUT and POST requests under rack
7b3b56e (5977) Puppet::Applications can be loaded from multiple paths.

2.6.5rc1
========
f9e2e2b Augmentation of tests for prior commit
392504a Fix to fix for #5755 -- backref serialization issues in zaml
a732a15 Fixed #5564 - Added some more fqdn_rand documentation
f279f2c Fixed #4968 - Updated list of options turned on by --test in documentation
ce5a2bf (#5061) - allow special hostclass/define variables to be evaluated as defaults.
fd73874 (#6107) Fix an error when auditing a file with empty content
530496b Remove already initialized constant warning from file_spec.rb tests
76788f8 (#5566) Treat source only File checksums as syntax errors when used with content
d657292 Rename variable used in File type validation to be more clear
3398139 Remove invalid "timestamp" and "time", and add missing "ctime" File checksum types.
6c93eb2 Remove order dependency when specifying source and checksum on File type
3a125d4 Bug #5755 -- ZAML generates extra newline in some hash backreferences.
50c12e5 bug #5681 -- code fix to handle AIX mount output
139760b Bug #5681 -- parse AIX mount command output.
2f74d83 Spec for #5681 to allow parsing of AIX mount output in mount provider
878f266 Fixed #6091 - Changed POSIX path matching to allow multiple leading slashes
eb97aa5 Bug #6091 -- test leading double-slash in filenames are allowed.
1bfc9a0 Fixed #6071 - Fixed typo and improved exec path error message
c50a48e Fixed #6061 - Allowed -1 as password min/max age
bf44e72 Bug #6061 -- verify that negative {min,max}_password_age are accepted.
af1c1fe Feature #5855 -- fix withenv call in freebsd package provider
d871641 Feature #5855 -- undefined method 'withenv' in FreeBSD package provider.
f1ab588 Fixed #6009 - nested member list vs directory service group provider
86a2a00 (#5944) Remove documentation of define() when used on nodes, as it is not a supported use of this function.
2b9f653 (#5944) Further edits of inline defined() documentation.
5d108e8 (#5944) Improve documentation of defined() function
7d38ab2 (#5594) Update documentation of exec resource type.
67e1bba (#5931) Prevent errors when calling insync? on audited properties
0f9d236 Maint: Removed dead code from resource harness.
0765afb Maint: Rename misleading insync? method in file provider
0084b08 (#5548) Specify return values of manual status commands in service type description.
dd332f6 Fixed #6002 - Added note about function execution
3cfbd07 (#5045) Cleaning up some tests and code
a2036ea (#5045) External node classifiers should be able to specify params for classes
18ca97b (#5045) Adds support to resource/type to also accept a param hash
70630b9 Fix #3165 Ralsh (bin/puppet resource) can't manage files
1fd3600 Fixed #3646 - Added documentation for compile and apply to man page
ae48634 Fixed #5914 Removed genconfig = true from genconfig output
7e7f342 Fixed #1657 - Added note about target file
069f29b Fixed #2096 - clarified option modification and tested it is working
66b442b Fixes #5916 - Cleanup of unused doc methods and documentation
9b74968 Modified rubydoc in lib/puppet/util/command_line/puppetca to fix inaccurate description of --clean.
e58f5dc Fixed #5742 - Removed legacy fqdn option from documentation
4d1b51f Fixed #5167 - misleading documentation in the defaults of [main]
c1b5c7f (#5913) Fix Puppet::Application.find constant lookup behavior
f9bfb96 (#5900) Include ResourceStatus#failed in serialized reports
79b6332 (#5882) Added error-handling for bucketing files in puppet inspect
17843d5 (#5882) Added error-handling to puppet inspect when auditing
1a6fab2 (#5171) Made "puppet inspect" upload audited files to a file bucket
a7cd185 Prep for #5171: Added a missing require to inspect application.
71ac9cf Locked Puppet license to GPLv2
abc6256 (#5838) Support paths as part of file bucket requests.
002f9f1 (#5838) Improve the quality of file bucket specs.
94d7179 (#5838) Make file bucket dipper efficient when saving a file that already exists
89f5692 (#5838) Implemented the "head" method for FileBucketFile::File terminus.
9cfd3d5 (#5838) Reworked file dipper spec to perform less stubbing.
c514c64 (#5838) Added support for HEAD requests to the indirector.
2b9b7a5 (#5838) Refactored error handling logic into find_in_cache.
08561b2 (#5838) Refactored Puppet::Network::Rights#fail_on_deny
87c5c30 (#5910) Improved logging when declared classes cannot be found:
4efc98a maint: Remove unused Rakefile in spec directory
a002231 (#5171) Made filebucket able to perform diffs
8f314f2 (#5710) Removed unnecessary calls to insync?
e270086 Prep for fixing #5710: Refactor stub provider in resource harness spec
c57a677 Maint: test partial resource failure
8aa8b9d (#5799) Simplify report dir creation
2d88844 maint: Add vim swap files to .gitignore
3d3baec maint: Remove rspec options from the Rakefile
df65304 maint: Inspect reports should have audited = true on events
4c9eca1 Maint: Added "skipped" to the YAML output for Puppet::Resource::Status
717670f (#5771): Fix spec failures associated with rspec upgrade
52760a4 (#5771) Upgrade rspec to version 2
7603b05 maint: remove stray debug statement.
7661ba8 maint: Prune #inspect methods on various objects
80bfb54 (#5758) Verify that report events are correctly created
de85f8d Prep work for #5758: set audited=true on all audit events

lib/puppet/application/pi.rb
require 'puppet'
require 'puppet/application'

class Formatter

    def initialize(width)
        @width = width
    end

    def wrap(txt, opts)
        return "" unless txt && !txt.empty?
        work = (opts[:scrub] ? scrub(txt) : txt)
        indent = (opts[:indent] ? opts[:indent] : 0)
        textLen = @width - indent
        patt = Regexp.new("^(.{0,#{textLen}})[ \n]")
        prefix = " " * indent

        res = []

        while work.length > textLen
            if work =~ patt
                res << $1
                work.slice!(0, $&.length)
            else
                res << work.slice!(0, textLen)
            end
        end
        res << work if work.length.nonzero?
        return prefix + res.join("\n" + prefix)
    end

    def header(txt, sep = "-")
        "\n#{txt}\n" + sep * txt.size
    end

    private

    def scrub(text)
        # For text with no carriage returns, there's nothing to do.
        if text !~ /\n/
            return text
        end
        indent = nil

        # If we can match an indentation, then just remove that same level of
        # indent from every line.
        if text =~ /^(\s+)/
            indent = $1
            return text.gsub(/^#{indent}/,'')
        else
            return text
        end
    end

end

class TypeDoc

    def initialize
        @format = Formatter.new(76)
        @types = {}
        Puppet::Type.loadall
        Puppet::Type.eachtype { |type|
            next if type.name == :component
            @types[type.name] = type
        }
    end

    def list_types
        puts "These are the types known to puppet:\n"
        @types.keys.sort { |a, b|
            a.to_s <=> b.to_s
        }.each do |name|
            type = @types[name]
            s = type.doc.gsub(/\s+/, " ")
            n = s.index(".")
            if n.nil?
                s = ".. no documentation .."
            elsif n > 45
                s = s[0, 45] + " ..."
            else
                s = s[0, n]
            end
            printf "%-15s - %s\n", name, s
        end
    end

    def format_type(name, opts)
        name = name.to_sym
        unless @types.has_key?(name)
            puts "Unknown type #{name}"
            return
        end
        type = @types[name]
        puts @format.header(name.to_s, "=")
        puts @format.wrap(type.doc, :indent => 0, :scrub => true) + "\n\n"

        puts @format.header("Parameters")
        if opts[:parameters]
            format_attrs(type, [:property, :param])
        else
            list_attrs(type, [:property, :param])
        end

        if opts[:meta]
            puts @format.header("Meta Parameters")
            if opts[:parameters]
                format_attrs(type, [:meta])
            else
                list_attrs(type, [:meta])
            end
        end

        if type.providers.size > 0
            puts @format.header("Providers")
            if opts[:providers]
                format_providers(type)
            else
                list_providers(type)
            end
        end
    end

    # List details about attributes
    def format_attrs(type, attrs)
        docs = {}
        type.allattrs.each do |name|
            kind = type.attrtype(name)
            if attrs.include?(kind) && name != :provider
                docs[name] = type.attrclass(name).doc
            end
        end

        docs.sort { |a,b|
            a[0].to_s <=> b[0].to_s
        }.each { |name, doc|
            print "\n- **%s**" % name
            if type.namevar == name and name != :name
                puts " (*namevar*)"
            else
                puts ""
            end
            puts @format.wrap(doc, :indent => 4, :scrub => true)
        }
    end

    # List the names of attributes
    def list_attrs(type, attrs)
        params = []
        type.allattrs.each do |name|
            kind = type.attrtype(name)
            if attrs.include?(kind) && name != :provider
                params << name.to_s
            end
        end
        puts @format.wrap(params.sort.join(", "), :indent => 4)
    end

    def format_providers(type)
        type.providers.sort { |a,b|
            a.to_s <=> b.to_s
        }.each { |prov|
            puts "\n- **%s**" % prov
            puts @format.wrap(type.provider(prov).doc,
                              :indent => 4, :scrub => true)
        }
    end

    def list_providers(type)
        list = type.providers.sort { |a,b|
            a.to_s <=> b.to_s
        }.join(", ")
        puts @format.wrap(list, :indent => 4)
    end

end

Puppet::Application.new(:pi,"#{$0} [options] [type]") do

    should_not_parse_config

    option("--short", "-s", "Only list parameters without detail") do |arg|
        options[:parameters] = false
    end

    option("--providers","-p")
    option("--list", "-l")
    option("--meta","-m")

    preinit do
        options[:parameters] = true
    end

    command(:main) do
        doc = TypeDoc.new

        if options[:list]
            doc.list_types
        else
            options[:types].each { |name| doc.format_type(name, options) }
        end
    end

    setup do
        options[:types] = ARGV.dup
        unless options[:list] || options[:types].size > 0
            handle_help(nil)
        end
        if options[:list] && options[:types].size > 0
            $stderr.puts "Warning: ignoring types when listing all types"
        end
    end

end
